crates/biome_js_analyze/src/lint/nursery/no_secrets.rs

```diff
@@ -155,119 +223,284 @@ struct SensitivePattern {
+            let entropy =
+                calculate_entropy_with_case_and_classes(token, *entropy_threshold as f64, 15.0);
+            if (entropy as u16) > *entropy_threshold {
+                return Some("Detected high entropy string");
+            }
+        }
+    }
+    None
 }
 
-/// Inspired by https://github.com/nickdeis/eslint-plugin-no-secrets/blob/master/utils.js#L93
-/// Adapted from https://docs.rs/entropy/latest/src/entropy/lib.rs.html#14-33
-/// Calculates Shannon entropy to measure the randomness of data. High entropy values indicate potentially
-/// secret or sensitive information, as such data is typically more random and less predictable than regular text.
-/// Useful for detecting API keys, passwords, and other secrets within code or configuration files.
-fn calculate_shannon_entropy(data: &str) -> f64 {
+/*
+Uses Shannon Entropy as a base algorithm, then adds "boosts" for special patterns/occurrences.
+For example, Continuous mixed cases (lIkE tHiS) are more likely to contribute to a higher score than single cases.
+Symbols also contribute highly to secrets.
+
+TODO: This needs work. False positives/negatives are highlighted in valid.js and invalid.js.
+
+References:
+- ChatGPT chat: https://chatgpt.com/share/670370bf-3e18-8011-8454-f3bd01be0319
+- Original paper for Shannon Entropy: https://ieeexplore.ieee.org/abstract/document/6773024/
+*/
+fn calculate_entropy_with_case_and_classes(
+    data: &str,
+    base_threshold: f64,
+    scaling_factor: f64,
+) -> f64 {
     let mut freq = [0usize; 256];
     let len = data.len();
+
     for &byte in data.as_bytes() {
         freq[byte as usize] += 1;
     }
 
-    let mut entropy = 0.0;
+    let mut shannon_entropy = 0.0;
+    let mut letter_count = 0;
+    let mut uppercase_count = 0;
+    let mut lowercase_count = 0;
+    let mut digit_count = 0;
+    let mut symbol_count = 0;
+    let mut case_switches = 0;
+    let mut previous_char_was_upper = false;
+
     for count in freq.iter() {
         if *count > 0 {
             let p = *count as f64 / len as f64;
-            entropy -= p * p.log2();
+            shannon_entropy -= p * p.log2();
+        }
+    }
+
+    // Letter classification and case switching
+    for (i, c) in data.chars().enumerate() {
+        if c.is_ascii_alphabetic() {
+            letter_count += 1;
+            if c.is_uppercase() {
+                uppercase_count += 1;
+                if i > 0 && !previous_char_was_upper {
+                    case_switches += 1;
+                }
+                previous_char_was_upper = true;
+            } else {
+                lowercase_count += 1;
+                if i > 0 && previous_char_was_upper {
+                    case_switches += 1;
+                }
+                previous_char_was_upper = false;
+            }
+        } else if c.is_ascii_digit() {
+            digit_count += 1;
+        } else if !c.is_whitespace() {
+            symbol_count += 1;
```

Comment:
Since we check only ASCII stuff, we could directly iterate over bytes. This is more efficient because this avoids decoding UTF-8 codepooints.

```suggestion
    for (i, c) in data.bytes().enumerate() {
        if c.is_ascii_alphabetic() {
            letter_count += 1;
            if c.is_ascii_uppercase() {
                uppercase_count += 1;
                if i > 0 && !previous_char_was_upper {
                    case_switches += 1;
                }
                previous_char_was_upper = true;
            } else {
                lowercase_count += 1;
                if i > 0 && previous_char_was_upper {
                    case_switches += 1;
                }
                previous_char_was_upper = false;
            }
        } else if c.is_ascii_digit() {
            digit_count += 1;
        } else if !c.is_ascii_whitespace() {
            symbol_count += 1;
```